API.md

```diff
@@ -1,6 +1,96 @@
+
+If configured as a global middleware(see usage section), all request need to provide a valid token to access the endpoint.
+<br>
+If some endpoints should be excluded from this restriction, they need to be added to the `options.publicEndpoints` array to be whitelisted.
+
+If validation of the provided token fails the middleware rejects the request with status _401 UNAUTHORIZED_. <br>
+To overwrite this behavior a custom handler can be specified by passing in `options.onNotAuthenticatedHandler` (see [`onNotAuthenticatedHandler`](./src/types/AuthenticationMiddlewareOptions.ts)).
+
+* ⚠️&nbsp;&nbsp;While this middleware could also be configured per endpoint (i.e. `app.get(authenticationMiddleware(...), endpoint)` it is not recommended as using it as global middleware will force you into a whitelist setup.
+  * Make sure `authenticationMiddleware` is at the top of the registered request handlers. This is essential to gurantee the enforceability of the whitelist strategy.
+* ⚠️&nbsp;&nbsp;The middleware attaches metadata (scopes of the token) to the express request object. The `requireScopesMiddleware` relies on this information.
+
+
+#### Usage
+
+```typescript
+import {
+  authenticationMiddleware
+} from 'authmosphere';
+
+app.use(authenticationMiddleware({
+  publicEndpoints: ['/heartbeat', '/status'],
+  tokenInfoEndpoint: 'auth.example.com/tokeninfo'
+});
+```
+
+#### Signature
+
+`authenticationMiddleware(options) => express.RequestHandler`
+
+#### Arguments
+
+* [`options`](./src/types/AuthenticationMiddlewareOptions.ts):
+  * `tokenInfoEndpoint: string` - url of the Token validation endpoint
+  * `publicEndpoints?: string[] - list of whitelisted API paths`
+  * `logger?: Logger` - [logger](./src/types/Logger.ts)
+  * [`onNotAuthenticatedHandler?: onNotAuthenticatedHandler`](./src/types/AuthenticationMiddlewareOptions.ts) - custom response handler
+
+### requireScopesMiddleware
+
+A factory that returns a middleware that compare scopes of the request token with a given list and either call `next` or reject request with 403 Forbidden.
+
+More complex description of the middleware's authorization flow:<br>
+  * if precedence function resolves
+    *  give control to next request handler (i.e. call `next()`)
+  * if precedenceFunction rejects and
+    * if token scopes from the request has at least the scopes specified in the `scopes` parameter
+      * give control to next request handler (i.e. call `next()`)
+    * if token scopes from the request has less scopes then specified in the `scopes` parameter
+      * __reject request__
+        * if custom handler is configured call: `options.onAuthorizationFailedHandler`
+        * if no custom handler is configured, respond with status Forbidden(403)
+
+* ⚠️&nbsp;&nbsp;This middleware requires that scope information is attached to the `request` object. The `authenticationMiddleware` can do this job. Otherwise `req.$$tokeninfo.scope: string[]` has to be set manually.
```

Comment:
Would move that before the "complex description" part.

Reqording suggestion:
>⚠️&nbsp;&nbsp;This middleware requires scope information to be attached to the `Express.request` object. The `authenticationMiddleware` can do this job. Otherwise `request.$$tokeninfo.scope: string[]` has to be set manually.

Reply:
✅